flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.java

```diff
@@ -0,0 +1,221 @@
+		inputState = getRuntimeContext().getMapState(mapStateDescriptor);
+
+		BaseRowTypeInfo accTypeInfo = new BaseRowTypeInfo(accTypes);
+		ValueStateDescriptor<BaseRow> stateDescriptor =
+			new ValueStateDescriptor<BaseRow>("accState", accTypeInfo);
+		accState = getRuntimeContext().getState(stateDescriptor);
+
+		initCleanupTimeState("ProcTimeBoundedRangeOverCleanupTime");
+	}
+
+	@Override
+	public void processElement(
+		BaseRow input,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.Context ctx,
+		Collector<BaseRow> out) throws Exception {
+		Long currentTime = ctx.timerService().currentProcessingTime();
+		// register state-cleanup timer
+		registerProcessingCleanupTimer(ctx, currentTime);
+
+		// buffer the event incoming event
+
+		// add current element to the window list of elements with corresponding timestamp
+		List<BaseRow> rowList = inputState.get(currentTime);
+		// null value means that this si the first event received for this timestamp
+		if (rowList == null) {
+			rowList = new ArrayList<BaseRow>();
+			// register timer to process event once the current millisecond passed
+			ctx.timerService().registerProcessingTimeTimer(currentTime + 1);
+		}
+		rowList.add(input);
+		inputState.put(currentTime, rowList);
+	}
+
+	@Override
+	public void onTimer(
+		long timestamp,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.OnTimerContext ctx,
+		Collector<BaseRow> out) throws Exception {
+		if (needToCleanupState(timestamp)) {
+			// clean up and return
+			cleanupState(inputState, accState);
+			function.cleanup();
+			return;
+		}
+
+		// remove timestamp set outside of ProcessFunction.
+		((TimestampedCollector) out).eraseTimestamp();
+
+		// we consider the original timestamp of events
+		// that have registered this time trigger 1 ms ago
+
+		long currentTime = timestamp - 1;
+		// get the list of elements of current proctime
+		List<BaseRow> currentElements = inputState.get(currentTime);
+
+		// Expired clean-up timers pass the needToCleanupState check.
+		// Perform a null check to verify that we have data to process.
+		if (null == currentElements) {
+			return;
+		}
+
+		// initialize the accumulators
+		BaseRow accumulators = accState.value();
+
+		if (null == accumulators) {
+			accumulators = function.createAccumulators();
+		}
+
+//		// set accumulators in context first
+//		function.setAccumulators(accumulators);
+
+		// update the elements to be removed and retract them from aggregators
+		long limit = currentTime - precedingTimeBoundary;
+
+		// we iterate through all elements in the window buffer based on timestamp keys
+		// when we find timestamps that are out of interest, we retrieve corresponding elements
+		// and eliminate them. Multiple elements could have been received at the same timestamp
+		// the removal of old elements happens only once per proctime as onTimer is called only once
+		Iterator<Map.Entry<Long, List<BaseRow>>> iter = inputState.iterator();
```

Comment:
Please iterate on the keys `inputState.keys()`. Iterate on all the entries will introduce a lot of unnecessary cost.